extra/opensslpp/include/opensslpp/core_error.hpp

```diff
@@ -0,0 +1,37 @@
+/* Copyright (c) 2022 Percona LLC and/or its affiliates. All rights reserved.
+
+   This program is free software; you can redistribute it and/or
+   modify it under the terms of the GNU General Public License
+   as published by the Free Software Foundation; version 2 of
+   the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA */
+
+#ifndef OPENSSLPP_CORE_ERROR_HPP
+#define OPENSSLPP_CORE_ERROR_HPP
+
+#include <stdexcept>
+#include <string>
+
+namespace opensslpp {
+
+class core_error : public std::runtime_error {
```

Comment:
This was my initial idea as well, but it turned out that for most cases providing OpenSSL error details to the user is very confusing - they are too low level, so I decided to fall back to more simple error messages in most cases.
But this can definitely be revised if you think some case may benefit from more info.
Probably not 100% true, but I do `raise_with_error_string()` mostly for PEM read/write functions - in this case OpenSSL indeed gives more details on what went wrong.

Reply:
> to the user is very confusing

`opensslpp` looks like a generic C++ OpenSSL wrapper library used in other projects, such as the UDF component, or other components/things in the future. And it's strange for a wrapper library to hide error messages from the underlying library.

On the other hand, I didn't check how those error messages would look like, I only checked the documentation of a few OpenSSL functions where these exceptions are generated, and most said that for more details about the error, use `err_get_error`.

Or these error messages might be confusing to users, but maybe they could be helpful for the next developer working on some code with `opensslpp`?

In that case, maybe `core_error` could store the real OpenSSL error in another data member with a separate accessor. The using code could decide if it wanted to display it to the user or not, and while debugging, it would be easily available to the developer?

Reply:
@dutow My judging is based on the following example.
If, for instance, we have an error in the opensslpp::rsa_key copy constructor (most probably one of the memory allocation functions failed during RSA components (big numbers, BNs) cloning), we do not need nothing more than "Cannot duplicate rsa_key". Extracting SSL error message won't give users any useful info in this case - the error would look like "Failed to allocate memory for BN_copy()" (users won't even know what BN in this case is).
On the other hand from my experience almost everything PEM format-related has pretty good/useful error diagnostics